### ISSUE TITLE
referer is undefined if no session

### ISSUE THREAD

**Issue by miyconst** [PuppetJs/PuppetJs#51]
If I return a page without session, then the following exception occurs in `puppet.js`
```javascript
Uncaught TypeError: Cannot read property 'replace' of null
```
```
puppet.js:191
```
```javascript
var wsPath = this.referer.replace(/__([^\/]*)\//g, "__$1/wsupgrade/");
```

**Comment by miyconst** [comment]
@warpech should `puppet.js` work without session or this is desired behaviour?

**Comment by warpech** [comment]
In my opinion `puppet.js` must have a session. @tomalec do you agree?

There is another issue regarding that the error message in this case is too cryptical: https://github.com/PuppetJs/PuppetJs/issues/49

**Comment by warpech** [comment]
Ping @tomalec 

Also, this is related to https://github.com/PuppetJs/PuppetJs/issues/58

**Comment by tomalec** [comment]
Technically I would say that `PuppetJs` itself does not need a session. It just needs (http/ws) URL to send patches to. So it's Starcounter server-side that needs a session. Solving it as in #58 would be cool, and simplify unclear presence of `this.referer` in puppet.

Until #58 gets clarified we canprovide nice error message.

**Comment by warpech** [comment]
After giving it a second thought, I agree with @tomalec. 

The JavaScript error in topic will be automatically gone after #58 is solved, because #58 merges `this.remoteUrl` and `this.referer`. 

The problem of what happens if you don't create Starcounter `Session` on server will remain though and let's discuss that one in issue #49.

**Comment by miyconst** [comment]
Should I close this issue then?

**Comment by warpech** [comment]
Please close it after #58 is solved

**Comment by warpech** [comment]
#58 is solved now. @miyconst can you check what does happen now if the server returns a page without a session?

**Comment by miyconst** [comment]
Server does not return `x-location` header/cookie if there is no session, that is why puppet tries to open ws connection to the initial `remoteUrl`, which is equal to `windlow.location.href`, and then fails with an exception:

```
WebSocket connection to 'ws://localhost:8080/images' failed: Error during WebSocket handshake: Unexpected response code: 200
```

P.S. I was testing on Images app.

**Comment by warpech** [comment]
I guess this error comes from the web browser, not from PuppetJs. Is this a fatal condition for PuppetJs or just a warning, after which PuppetJs works in HTTP PATCH mode?

If this is not a fatal condition for the client then I think the problem is solved. The programmer should understand from the error message that he needs to set `useWebSockets` to false on client, or implement WS on server.

**Comment by miyconst** [comment]
This is an error from PuppetJs:

```js
that._ws.onerror = function (event) {
   throw new Error(" ... ");
}
```

App in such state is not usable. The servers responds with `406` to `GET application/json-patch+json` requests and `404` to `PATCH application/json-patch+json` requests.

**Comment by warpech** [comment]
I think this is OK for the client. 

I find the 404 response to be misleading and that should be fixed on the server side. I will create a separate issue about that.

**Comment by warpech** [comment]
Closing as fixed. Changing the misleading server responses is tracked in #49.
